consensus/storage_fault_monitor.go

```diff
@@ -0,0 +1,89 @@
+package consensus
+
+import (
+	"context"
+
+	"github.com/ipfs/go-cid"
+	logging "github.com/ipfs/go-log"
+
+	"github.com/filecoin-project/go-filecoin/abi"
+	"github.com/filecoin-project/go-filecoin/address"
+	"github.com/filecoin-project/go-filecoin/types"
+	"github.com/filecoin-project/go-filecoin/vm/errors"
+)
+
+// monitorPlumbing is an interface for the functionality StorageFaultSlasher needs
+type monitorPlumbing interface {
+	MessageQuery(ctx context.Context, optFrom, to address.Address, method string, params ...interface{}) ([][]byte, error)
+}
+
+// slashingMsgOutbox is the interface for the functionality of Outbox StorageFaultSlasher needs
+type slashingMsgOutbox interface {
+	Send(ctx context.Context,
+		from, to address.Address,
+		value types.AttoFIL,
+		gasPrice types.AttoFIL,
+		gasLimit types.GasUnits,
+		bcast bool,
+		method string,
+		params ...interface{}) (out cid.Cid, err error)
+}
+
+// StorageFaultSlasher checks for unreported storage faults by miners, a.k.a. market faults.
+// Storage faults are distinct from consensus faults.
+// See https://github.com/filecoin-project/specs/blob/master/faults.md
+type StorageFaultSlasher struct {
+	log       logging.EventLogger
+	msgSender address.Address   // what signs the slashing message and receives slashing reward
+	outbox    slashingMsgOutbox // what sends the slashing message
+	plumbing  monitorPlumbing   // what does the message query
+}
+
+// NewStorageFaultMonitor creates a new StorageFaultSlasher with the provided plumbing, outbox and message sender.
+// Message sender must be an account actor address.
+func NewStorageFaultMonitor(plumbing monitorPlumbing, outbox slashingMsgOutbox, msgSender address.Address) *StorageFaultSlasher {
+	return &StorageFaultSlasher{
+		plumbing:  plumbing,
+		log:       logging.Logger("StorFltMon"),
+		outbox:    outbox,
+		msgSender: msgSender,
+	}
+}
+
+// Slash checks for miners with unreported faults, then slashes them
+// Slashing messages are not broadcast to the network, but included in the next block mined by the slashing
+// node.
+func (sfm *StorageFaultSlasher) Slash(ctx context.Context, currentHeight *types.BlockHeight) error {
+	res, err := sfm.plumbing.MessageQuery(ctx, sfm.msgSender, address.StorageMarketAddress, "getLateMiners")
+	if err != nil {
+		return errors.FaultErrorWrap(err, "getLateMiners message failed")
```

Comment:
nit: Are fault errors the right thing to return right here?  My impression is that they are used internally within state machine processing but not useful for querying.

Reply:
@acruikshank  ? this was a request from you I think

Reply:
I can't remember the context under which I would have said that. At any rate, Fault and Revert errors are meaningless outside of message execution, so wrapping the error in a Fault doesn't help and doesn't hurt. It might be better to wrap it some other way to avoid the type of confusion we're experiencing now.

Reply:
I have changes for this file anyway and I'll just convert it to a plain wrap